File: src/main/java/play/sockjs/SockJS.java
package play.sockjs;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;

import akka.stream.javadsl.Flow;

import com.fasterxml.jackson.databind.JsonNode;
import play.api.http.websocket.CloseCodes;
import play.libs.F;
import play.libs.Scala;
import play.libs.streams.AkkaStreams;
import play.mvc.*;
import scala.PartialFunction;

public abstract class SockJS {

    /**
     * Invoke the SockJS.
     *
     * @param request The request for the SockJS handler.
     * @return A future of either a result to reject the SockJS connection with, or a Flow to handle the SockJS.
     */
    public abstract CompletionStage<F.Either<Result, Flow<Frame, Frame, ?>>> apply(Http.RequestHeader request);

    /**
     * Acceptor for text SockJS.
     */
    public static final MappedSockJSAcceptor<String, String> Text = new MappedSockJSAcceptor<>(Scala.partialFunction(message -> {
        if (message instanceof Frame.Text) {
            return F.Either.Left(((Frame.Text) message).data());
        } else {
            throw Scala.noMatch();
        }
    }), Frame.Text::new);

    /**
     * Acceptor for JSON SockJS.
     */
    public static final MappedSockJSAcceptor<JsonNode, JsonNode> Json = new MappedSockJSAcceptor<>(Scala.partialFunction(message -> {
        try {
            if (message instanceof Frame.Text) {
                return F.Either.Left(play.libs.Json.parse(((Frame.Text) message).data()));
            }
        } catch (RuntimeException e) {
            return F.Either.Right(new Frame.Close(CloseCodes.Unacceptable(), "Unable to parse JSON message"));
        }
        throw Scala.noMatch();
    }), json -> new Frame.Text(play.libs.Json.stringify(json)));

    /**
     * Acceptor for JSON SockJS.
     *
     * @param in The class of the incoming messages, used to decode them from the JSON.
     * @param <In> The SockJS's input type (what it receives from clients)
     * @param <Out> The SockJS's output type (what it writes to clients)
     * @return The SockJS acceptor.
     */
    public static <In, Out> MappedSockJSAcceptor<In, Out> json(Class<In> in) {
        return new MappedSockJSAcceptor<>(Scala.partialFunction(message -> {
            try {
                if (message instanceof Frame.Text) {
                    return F.Either.Left(play.libs.Json.mapper().readValue(((Frame.Text) message).data(), in));
                }
            } catch (Exception e) {
                return F.Either.Right(new Frame.Close(CloseCodes.Unacceptable(), e.getMessage()));
            }
            throw Scala.noMatch();
        }), outMessage -> {
            try {
                return new Frame.Text(play.libs.Json.mapper().writeValueAsString(outMessage));
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        });
    }

    /**
     * Utility class for creating SockJS.
     *
     * @param <In> the type the SockJS reads from clients (e.g. String, JsonNode)
     * @param <Out> the type the SockJS outputs back to remote clients (e.g. String, JsonNode)
     */
    public static class MappedSockJSAcceptor<In, Out> {
        private final PartialFunction<Frame, F.Either<In, Frame>> inMapper;
        private final Function<Out, Frame> outMapper;

        public MappedSockJSAcceptor(PartialFunction<Frame, F.Either<In, Frame>> inMapper, Function<Out, Frame> outMapper) {
            this.inMapper = inMapper;
            this.outMapper = outMapper;
        }

        /**
         * Accept a SockJS.
         *
         * @param f A function that takes the request header, and returns a future of either the result to reject the
         *          SockJS connection with, or a flow to handle the SockJS messages.
         * @return The SockJS handler.
         */
        public SockJS acceptOrResult(Function<Http.RequestHeader, CompletionStage<F.Either<Result, Flow<In, Out, ? >>>> f) {
            return SockJS.acceptOrResult(inMapper, f, outMapper);
        }

        /**
         * Accept a SockJS.
         *
         * @param f A function that takes the request header, and returns a flow to handle the SockJS messages.
         * @return The SockJS handler.
         */
        public SockJS accept(Function<Http.RequestHeader, Flow<In, Out, ? >> f) {
            return acceptOrResult(request -> CompletableFuture.completedFuture(F.Either.Right(f.apply(request))));
        }
    }

    /**
     * Helper to create handlers for SockJS.
     *
     * @param inMapper Function to map input messages. If it produces left, the message will be passed to the SockJS
     *                 flow, if it produces right, the message will be sent back out to the client - this can be used
     *                 to send errors directly to the client.
     * @param f The function to handle the SockJS.
     * @param outMapper Function to map output messages.
     * @return The SockJS handler.
     */
    private static <In, Out> SockJS acceptOrResult(
            PartialFunction<Frame, F.Either<In, Frame>> inMapper,
            Function<Http.RequestHeader, CompletionStage<F.Either<Result, Flow<In, Out, ?>>>> f,
            Function<Out, Frame> outMapper
    ) {
        return new SockJS() {
            @Override
            public CompletionStage<F.Either<Result, Flow<Frame, Frame, ?>>> apply(Http.RequestHeader request) {
                return f.apply(request).thenApply(resultOrFlow -> {
                    if (resultOrFlow.left.isPresent()) {
                        return F.Either.Left(resultOrFlow.left.get());
                    } else {
                        Flow<Frame, Frame, ?> flow = AkkaStreams.bypassWith(
                                Flow.<Frame>create().collect(inMapper),
                                play.api.libs.streams.AkkaStreams.onlyFirstCanFinishMerge(2),
                                resultOrFlow.right.get().map(outMapper::apply)
                        );
                        return F.Either.Right(flow);
                    }
                });
            }
        };
    }
}

File: src/main/scala/play/sockjs/api/SockJS.scala
package play.sockjs.api

import scala.concurrent.Future
import scala.util.control.NonFatal

import akka.actor.{ActorRef, Props}
import akka.stream.scaladsl._

import play.api.http.websocket.CloseCodes
import play.api.libs.streams._
import play.api.libs.json._
import play.api.mvc._

import play.sockjs.api.Frame._

/**
  * A SockJS handler.
  */
trait SockJS extends Handler {

  /**
    * Execute the SockJS handler.
    *
    * The return value is either a result to reject the SockJS connection with,
    * or a flow that will handle the SockJS messages.
    */
  def apply(request: RequestHeader): Future[Either[Result, Flow[Frame, Frame, _]]]
}

object SockJS {

  def apply(f: RequestHeader => Future[Either[Result, Flow[Frame, Frame, _]]]): SockJS = {
    new SockJS {
      def apply(request: RequestHeader) = f(request)
    }
  }

  /**
    * Transforms SockJS message flows into flows of another type.
    *
    * The transformation may be more than just converting from one message to another, it may also produce messages, such
    * as close messages with an appropriate error code if the message can't be consumed.
    */
  trait MessageFlowTransformer[+In, -Out] { self =>

    /**
      * Transform the flow of In/Out messages into a flow of SockJS frames.
      */
    def transform(flow: Flow[In, Out, _]): Flow[Frame, Frame, _]

    /**
      * Contramap the out type of this transformer.
      */
    def contramap[NewOut](f: NewOut => Out): MessageFlowTransformer[In, NewOut] = {
      new MessageFlowTransformer[In, NewOut] {
        def transform(flow: Flow[In, NewOut, _]) = self.transform(flow map f)
      }
    }

    /**
      * Map the in type of this transformer.
      */
    def map[NewIn](f: In => NewIn): MessageFlowTransformer[NewIn, Out] = {
      new MessageFlowTransformer[NewIn, Out] {
        def transform(flow: Flow[NewIn, Out, _]) = self.transform(Flow[In] map f via flow)
      }
    }

    /**
      * Map the in type and contramap the out type of this transformer.
      */
    def map[NewIn, NewOut](f: In => NewIn, g: NewOut => Out): MessageFlowTransformer[NewIn, NewOut] = {
      new MessageFlowTransformer[NewIn, NewOut] {
        def transform(flow: Flow[NewIn, NewOut, _]) = self.transform(Flow[In] map f via flow map g)
      }
    }
  }

  /**
    * Defaults message transformers.
    */
  object MessageFlowTransformer {

    implicit val identityFrameFlowTransformer: MessageFlowTransformer[Frame, Frame] = {
      new MessageFlowTransformer[Frame, Frame] {
        def transform(flow: Flow[Frame, Frame, _]) = flow
      }
    }

    /**
      * Converts text message to/from Strings.
      */
    implicit val stringFrameFlowTransformer: MessageFlowTransformer[String, String] = {
      new MessageFlowTransformer[String, String] {
        def transform(flow: Flow[String, String, _]) = {
          AkkaStreams.bypassWith[Frame, Vector[String], Frame](Flow[Frame] collect {
            case Text(data) => Left(data)
          })(Flow[Vector[String]].mapConcat[String](identity) via flow map Text.apply)
        }
      }
    }

    /**
      * Converts messages to/from JsValue.
      */
    implicit val jsonMessageFlowTransformer: MessageFlowTransformer[JsValue, JsValue] = {
      def closeOnException[T](block: => T) = try {
        Left(block)
      } catch {
        case NonFatal(e) => Right(Frame.Close(CloseCodes.Unacceptable, "Unable to parse json message"))
      }

      new MessageFlowTransformer[JsValue, JsValue] {
        def transform(flow: Flow[JsValue, JsValue, _]) = {
          AkkaStreams.bypassWith[Frame, Vector[JsValue], Frame](Flow[Frame] collect {
            case Text(data) => closeOnException(data.map(Json.parse))
          })(Flow[Vector[JsValue]].mapConcat[JsValue](identity) via flow map (json => Text(Json.stringify(json))))
        }
      }
    }

    /**
      * Converts messages to/from a JSON high level object.
      *
      * If the input messages fail to be parsed, the SockJS will be closed with an 1003 close code and the parse error
      * serialised to JSON.
      */
    def jsonMessageFlowTransformer[In: Reads, Out: Writes]: MessageFlowTransformer[In, Out] = {
      jsonMessageFlowTransformer.map(json => Json.fromJson[In](json).fold(
        errors => throw SockJSCloseException(Close(CloseCodes.Unacceptable, Json.stringify(JsError.toJson(errors)))),
        identity
      ), out => Json.toJson(out))
    }
  }

  /**
    * Accepts a SockJS using the given flow.
    */
  def accept[In, Out](f: RequestHeader => Flow[In, Out, _])(implicit transformer: MessageFlowTransformer[In, Out]): SockJS = {
    acceptOrResult(f.andThen(flow => Future.successful(Right(flow))))
  }

  /**
    * Creates an action that will either accept the SockJS connection, using the given flow to handle the in and out stream, or
    * return a result to reject the request.
    */
  def acceptOrResult[In, Out](f: RequestHeader => Future[Either[Result, Flow[In, Out, _]]])(implicit transformer: MessageFlowTransformer[In, Out]): SockJS = {
    SockJS(request =>
      f(request).map(_.right.map(transformer.transform))(play.core.Execution.trampoline)
    )
  }

  /**
    * A function that, given an actor to send upstream messages to, returns actor props to create an actor to handle
    * the SockJS connection
    */
  type HandlerProps = ActorRef => Props
}